ramips: add support for H3C TX1800 Plus / TX1801 Plus / TX1806

DragonBluep opened the pull request:
```
H3C TX180x series WiFi6 routers are customized by different carrier.
While these three devices look different, they use the same motherboard
inside. Another minor difference comes from the model name definition
in the u-boot environment variable.

Specifications:
 SOC:      MT7621 + MT7915
 ROM:      128 MiB
 RAM:      256 MiB
 LED:      status *2
 Button:   reset *1 + wps/mesh *1
 Ethernet:        lan *3 + wan *1 (10/100/1000Mbps)
 TTL Baudrate:    115200
 TFTP server IP:  192.168.124.99

MAC Address:
 use        address(sample 1)   address(sample 2)    source
 label      88:xx:xx:98:xx:12   88:xx:xx:a2:xx:a5   u-boot-env@ethaddr
 lan        88:xx:xx:98:xx:13   88:xx:xx:a2:xx:a6   $label +1
 wan        88:xx:xx:98:xx:12   88:xx:xx:a2:xx:a5   $label
 WiFi4_2G   8a:xx:xx:58:xx:14   8a:xx:xx:52:xx:a7   (Compatibility mode)
 WiFi5_5G   8a:xx:xx:b8:xx:14   8a:xx:xx:b2:xx:a7   (Compatibility mode)
 WiFi6_2G   8a:xx:xx:18:xx:14   8a:xx:xx:12:xx:a7
 WiFi6_5G   8a:xx:xx:78:xx:14   8a:xx:xx:72:xx:a7

Compatibility mode is used to guarantee the connection of old devices
that only support WiFi4 or WiFi5.

TFTP + TTL Installation:
Although a TTL connection is required for installation, we do not need
to tear down it. We can find the TTL port from the cooling hole at the
bottom. It is located below LAN3 and the pins are defined as follows:
|LAN1|LAN2|LAN3|----|WAN|
--------------------
    |GND|TX|RX|VCC|

1. Set tftp server IP to 192.168.124.99 and put initramfs firmware in
   server's root directory, rename it to a simple name "initramfs.bin".
2. Plug in the power supply and wait for power on, connect the TTL cable
   and open a TTL session, enter "reboot", then enter "Y" to confirm.
   Finally push "0" to interruput boot while booting.
3. Execute command to install a initramfs system:
   # tftp 0x80010000 192.168.124.99:initramfs.bin
   # bootm 0x80010000
4. Backup nand flash by OpenWrt LuCI or dd instruction. We need those
   partitions if we want to back to stock firmwre due to official
   website does not provide download link.
   # dd if=/dev/mtd1 of=/tmp/u-boot-env.bin
   # dd if=/dev/mtd4 of=/tmp/firmware.bin
5. Edit u-boot env to ensure use default bootargs and first image slot:
   # fw_setenv bootargs
   # fw_setenv bootflag 0
6. Upgrade sysupgrade firmware.
7. About restore stock firmware: flash the "firmware" and "u-boot-env"
   partitions that we backed up in step 4.
   # mtd write /tmp/u-boot-env.bin u-boot-env
   # mtd write /tmp/firmware.bin firmware

Additional Info:
The H3C stock firmware has a 160-byte firmware header that appears to
use a non-standard CRC32 verification algorithm. For this part of the
data, the u-boot does not check it so we can just directly replace it
with a placeholder.
```
**Next Steps:**
- [x] Test the `2 Gbps CPU bandwidth #10238` feature (later when I have time) and wait until it is merged.
- [x] Add support for another variant `H3C TX1801 Plus`.

Test Branch: [DragonBluep:pr_add_support_for_H3C_TX1806_test](https://github.com/DragonBluep/openwrt/tree/pr_add_support_for_H3C_TX1806_test)

DragonBluep commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -630,6 +630,26 @@ define Device/gnubee_gb-pc2
+	fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb
+  IMAGES += factory.bin
+  IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
+  IMAGE/factory.bin := pad-extra 160 | append-kernel | pad-to $$(KERNEL_SIZE) | \
```
sysupgrade.bin not work due to u-boot load kernel from kenel@0xa0 (so that we need pad extra 160 bytes before kernel). What do I need to do to add offsets to sysupgrade.bin?

mans0n replied:
Hmm, how about padding the kernel directly (in `KERNEL`)?
BTW why does it need extra 160 bytes? Is it some kind of vendor-defined header or just blank spaces?

DragonBluep replied:
Yes, there should be an h3c header here, but it uses a non-standard crc32 algorithm, and I have no way of parsing it and no GPL code to refer to. This firmware header only affects upgrades from the web interface, u-boot does not verify this part.
H3C NX18 is a "SPI Flash Version" of H3C TX1806, just leave a firmware download link here, hope those who are experienced and interested can take a look.
[H3C NX18 Firmware](http://www.h3c.com/cn/d_202103/1389284_30005_0.htm)
some investigations:
0x00 ~ 0x07 fixed string "H3C_+k+f"
0x3c ~ 0x3f firmware version (time stamp or other, not sure)
0x44 ~ 0x47 symmetrical with “0x3c ~ 0x3f”
0x48 ~ 0x4f non-standard crc (kernel + rootfs I guess)
0x50 ~ 0x53 kernel size
0x54 ~ 0x57 rootfs size
0x62 ~ 0x70 firmware name (include board name)

DragonBluep commented on target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh:
```diff
@@ -33,6 +33,15 @@ platform_do_upgrade() {
+		bootargs="$(fw_printenv -n bootargs 2>/dev/null)"
+		if [ -n "$bootargs" ]; then
+			fw_setenv bakargs "$bootargs" && fw_setenv bootargs || exit 1
+		fi
```
`bootargs=console=ttyS0,115200 rootfstype=squashfs loglevel=3 root=/dev/mtdblock7/8/9/..`
rename bootargs to ensure kernel get the correct partition and full log support.

mans0n replied:
Err, can this be resolved by adding `bootargs-append = " root="` in the DTS "chosen" node?
Even if not, we should access UART during installation anyway, I believe it's better to handle it once in the U-Boot shell directly, rather than poking u-boot-env for every sysupgrade.

DragonBluep replied:
I updated the flash instructions and now we can drop it. When "bootargs" is removed u-boot will use default arguments and uart will work as usual.

mans0n commented:
> wifi5_2g  8a:xx:xx:58:xx:14
> wifi5_5g  8a:xx:xx:b8:xx:14
> wifi6_2g  8a:xx:xx:18:xx:14
> wifi6_5g  8a:xx:xx:78:xx:14

What's the difference between wifi5 and wifi6?

> Due to we do not know wifi mac calculation algorithm, so re-define
"$label +2" as wlan0 mac and "$label +3" as wlan1 mac.

The device owns label+2 address but it does not own label+3, I don't think that's a safe practice. That's why many devices chose to use its LAN address with local bit set.

mans0n commented on target/linux/ramips/dts/mt7621_h3c_tx1806.dts:
```diff
@@ -0,0 +1,167 @@
+			label = "amber:status";
+			color = <LED_COLOR_ID_AMBER>;
+			function = LED_FUNCTION_STATUS;
+			function-enumerator = <0>;
```
`function-enumerator` is used when there's multiple leds with same name (color:function). You don't need this here.

mans0n commented on target/linux/ramips/dts/mt7621_h3c_tx1806.dts:
```diff
@@ -0,0 +1,167 @@
+			};
+		};
+
+		/* This reserved partition is not defined on stock firmware. It's used
+		   to store some information about nand flash. Just list it here and we
+		   shouldn't touch it.
+		partition@7800000 {
+			label = "reserved";
+			reg = <0x7800000 0x800000>;
+			read-only;
+		};*/
```
Many MT7621 NAND devices have this reserved area but they just skip this partition. IMO just a brief note (without the partition node) is sufficient here.

DragonBluep committed:
ramips: add support for H3C TX1800 Plus / TX1801 Plus / TX1806

H3C TX180x series WiFi6 routers are customized by different carrier.
While these three devices look different, they use the same motherboard
inside. Another minor difference comes from the model name definition
in the u-boot environment variable.

Specifications:
 SOC:      MT7621 + MT7915
 ROM:      128 MiB
 RAM:      256 MiB
 LED:      status *2
 Button:   reset *1 + wps/mesh *1
 Ethernet:        lan *3 + wan *1 (10/100/1000Mbps)
 TTL Baudrate:    115200
 TFTP server IP:  192.168.124.99

MAC Address:
 use        address(sample 1)   address(sample 2)    source
 label      88:xx:xx:98:xx:12   88:xx:xx:a2:xx:a5   u-boot-env@ethaddr
 lan        88:xx:xx:98:xx:13   88:xx:xx:a2:xx:a6   $label +1
 wan        88:xx:xx:98:xx:12   88:xx:xx:a2:xx:a5   $label
 WiFi4_2G   8a:xx:xx:58:xx:14   8a:xx:xx:52:xx:a7   (Compatibility mode)
 WiFi5_5G   8a:xx:xx:b8:xx:14   8a:xx:xx:b2:xx:a7   (Compatibility mode)
 WiFi6_2G   8a:xx:xx:18:xx:14   8a:xx:xx:12:xx:a7
 WiFi6_5G   8a:xx:xx:78:xx:14   8a:xx:xx:72:xx:a7

Compatibility mode is used to guarantee the connection of old devices
that only support WiFi4 or WiFi5.

TFTP + TTL Installation:
Although a TTL connection is required for installation, we do not need
to tear down it. We can find the TTL port from the cooling hole at the
bottom. It is located below LAN3 and the pins are defined as follows:
|LAN1|LAN2|LAN3|----|WAN|
--------------------
    |GND|TX|RX|VCC|

1. Set tftp server IP to 192.168.124.99 and put initramfs firmware in
   server's root directory, rename it to a simple name "initramfs.bin".
2. Plug in the power supply and wait for power on, connect the TTL cable
   and open a TTL session, enter "reboot", then enter "Y" to confirm.
   Finally push "0" to interruput boot while booting.
3. Execute command to install a initramfs system:
   # tftp 0x80010000 192.168.124.99:initramfs.bin
   # bootm 0x80010000
4. Backup nand flash by OpenWrt LuCI or dd instruction. We need those
   partitions if we want to back to stock firmwre due to official
   website does not provide download link.
   # dd if=/dev/mtd1 of=/tmp/u-boot-env.bin
   # dd if=/dev/mtd4 of=/tmp/firmware.bin
5. Edit u-boot env to ensure use default bootargs and first image slot:
   # fw_setenv bootargs
   # fw_setenv bootflag 0
6. Upgrade sysupgrade firmware.
7. About restore stock firmware: flash the "firmware" and "u-boot-env"
   partitions that we backed up in step 4.
   # mtd write /tmp/u-boot-env.bin u-boot-env
   # mtd write /tmp/firmware.bin firmware

Additional Info:
The H3C stock firmware has a 160-byte firmware header that appears to
use a non-standard CRC32 verification algorithm. For this part of the
data, the u-boot does not check it so we can just directly replace it
with a placeholder.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep commented:
> What's the difference between wifi5 and wifi6?

WiFi5 only supports 802.11ac, which is a compatibility mode. Compatibility mode is used to guarantee the connection of old devices and H3C TX1806 has only one mt7915 chip.

> The device owns label+2 address but it does not own label+3, I don't think that's a safe practice. That's why many devices chose to use its LAN address with local bit set.

Yes, you are right, now the mac address is the same as the factory value, but I'm not sure if this will work for other devices.

DragonBluep commented:
Refer to 275a76e59a37114dce4fdedbadabd73d5558abc8, disable background-radar due to there is no background-radar chain (left top) antenna connected to it.
![IMG20220529121954](https://user-images.githubusercontent.com/70847398/170852196-07ee8272-b862-484c-8be3-50906c17aae5.jpg)

DragonBluep commented:
Wait for mediatek NMBM support

mans0n commented:
I thought it was already added by 06382d1af7b2, though I haven't tested it myself. Isn't it working well enough?

DragonBluep commented:
@mans0n It looks like everything is working fine. [dmesg_202206191619.txt](https://github.com/openwrt/openwrt/files/8934974/dmesg_202206191619.txt)
`"mediatek,bmt-force-create"` is a little dangerous so I do not enable it. It appears to be used to create NMBM tables, but this should be done by u-boot. I have a question, does ubi have its own bad block management function, it seems that we do not need to protect the ubi partition.

mans0n commented:
Thanks for the reference!

> I have a question, does ubi have its own bad block management function, it seems that we do not need to protect the ubi partition.

Yeah, I heard that we should disable bmt over the ubi blocks, that's why b86452f8414d introduced `mediatek,bmt-remap-range`.

DragonBluep commented:
changes:
Add the 2 Gbps bandwidth feature #10238. **This commit based on it.**
Drop `mediatek,empty-page-ecc-protected;` seems not very useful.
Update commit message, now we do not need tear down it.
Add support for H3C TX1800 / TX1801 Plus. They look very different, so I didn't use the `DEVICE_ALT_MODEL` method.
[TX1806](http://www.h3c.com/cn/Products___Technology/IntelligentTerminalProducts/Standard-Network/T/H3C_TX1806/) 
[TX1801 Plus](http://www.h3c.com/cn/Products___Technology/IntelligentTerminalProducts/Standard-Network/T/H3C_TX1801_Plus/) 
[TX1800 Plus](http://www.h3c.com/cn/Products___Technology/IntelligentTerminalProducts/Standard-Network/T/H3C_TX1800_Plus/)

hauke commented:
Thank you for the patch, I applied it tot master in 13308161788c98ae6cd48c22b13339fdb8c77130.

riobard commented:
@DragonBluep Thanks very much for the hard work! 👍

I wonder if there's further work needed to make the installation process simpler for end-users? As of now the [commit](https://github.com/openwrt/openwrt/pull/9488) suggests installation is via TTL only. Is it possible to install via the stock firmware's web interface?

DragonBluep commented:
@riobard The official firmware will verify the firmware header. But I haven't found the GPL code of H3C, and I don't have enough experience to parse the firmware header, so this is almost impossible to build a web install image for me. 

Known firmware header information:
0x00 ~ 0x07 fixed string "H3C_+k+f"
0x3c ~ 0x3f firmware version (time stamp or other, not sure)
0x44 ~ 0x47 symmetrical with “0x3c ~ 0x3f”
0x48 ~ 0x4f non-standard crc (kernel + rootfs I guess)
0x50 ~ 0x53 kernel size
0x54 ~ 0x57 rootfs size
0x62 ~ 0x70 firmware name (include board name)

At least now we don't need to tear down the router. 😂
![image](https://user-images.githubusercontent.com/70847398/197224287-00774a33-0059-4e27-ac9c-7f37b102e888.png)

riobard commented:
@DragonBluep Thanks for the clarification! Guess I'll need to figure out the TTL installation process to give it a try 😄

BTW I have an unreleased firmware (`TX1801PCV100R008.bin` to be precise) for TX1801 Plus from H3C technical support to fix IPv6 issues. Not sure if it is of any help but if you're interested to take a look, I can send it to you. (GitHub does not allow me to paste the .bin file directly here for unknown reasons, even after gzip 🤷‍♂️).

DragonBluep commented:
> BTW I have an unreleased firmware (`TX1801PCV100R008.bin` to be precise) for TX1801 Plus from H3C technical support to fix IPv6 issues. Not sure if it is of any help but if you're interested to take a look, I can send it to you.

It may be helpful to other professionals. I have already extracted some old firmware: https://github.com/DragonBluep/archive/tree/main/H3C%20TX180X/Stock%20Firmware

DragonBluep commented:
I made a mistake before. The OpenWrt official initramfs image does not include `uboot-envtools`. We can download the initramfs-kernel.bin here as a workaround before https://github.com/openwrt/openwrt/pull/10368 is merged:
https://github.com/DragonBluep/openwrt/actions/runs/3197141642

riobard commented:
@DragonBluep Will the three models of H3C devices (HX18xx) appear in the official OpenWRT [Table of Hardware](https://openwrt.org/toh/start)? There's no guide/device page for these supported devices. Please let me know if there's anything I can help to create those missing pieces.

DragonBluep commented:
It seems that every once in a while, the administrator will add wiki pages for these newly added devices. But we don't have permission to edit it. In fact, I don't know how to add it manually.
I created a topic in this forum, but it is written in Chinese: https://www.right.com.cn/forum/thread-8244730-1-1.html

DragonBluep commented:
@riobard Oh I see it. If you are interested, maybe you can check how to add it. https://forum.openwrt.org/t/applying-for-openwrt-wiki-account/101671

riobard commented:
@DragonBluep Just requested a WiKi account. Let's see if it gets approved. 

Meanwhile, I took a look at the bottom of my H3C TX1801 Plus router. Under the ventilation mesh I see four pads at the location shown [here](https://github.com/openwrt/openwrt/pull/9488#issuecomment-1287072487) but the pads don't seem to have any holes. Does it mean I have to tear apart the device and weld TTL cables to the pads? 

![F59730C5-7DF3-4D28-9A2B-59BEA60EB172_1_201_a](https://user-images.githubusercontent.com/22984/200496805-e8810772-d4c4-41a7-bced-6c0cf85b99bb.jpeg)

DragonBluep commented:
@riobard You only need to use three long probes and press them with your hands. It would be very easy if someone could help you.
![image](https://user-images.githubusercontent.com/70847398/200503426-dc5ddff6-94de-422f-b93a-b353c6e537a3.png)

riobard commented:
@DragonBluep Thanks! So just maintaining firm contact with the pads is enough? Sorry for the precautions since I've never used TTL cables before 😛

DragonBluep commented:
@riobard Yes. Don't worry, even if something goes wrong, it will not broke your router. You just need to try again.
